consume panic

#### Expected behavior

no panic

#### Actual behavior
panic

#### System configuration
**Pulsar version**: v0.7.0

#### panic log
```
m_xx subscription=qwe topic="persistent://pulsar-asdasdad/formal/qweqwe-partition-3"
panic: runtime error: slice bounds out of range [:277] with capacity 273

goroutine 598 [running]:
github.com/apache/pulsar-client-go/pulsar/internal.(*buffer).Read(0xc0050dc8c0, 0x106, 0x10d, 0x1e40040, 0xc0050b9170)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/buffer.go:113 +0x6a
github.com/apache/pulsar-client-go/pulsar/internal.(*MessageReader).readSingleMessage(0xc004e1cc10, 0xc000000111, 0x45b17c, 0x0, 0x19d7e60, 0xc0050da401, 0x2fbbfc0)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/commands.go:150 +0xed
github.com/apache/pulsar-client-go/pulsar/internal.(*MessageReader).ReadMessage(0xc004e1cc10, 0x3ff0000000000000, 0x0, 0x2fbbfc0, 0x1e71520, 0xc0050dc8c0, 0x0)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/commands.go:130 +0x60
github.com/apache/pulsar-client-go/pulsar.(*partitionConsumer).MessageReceived(0xc004cd6680, 0xc0050b1360, 0x1e71520, 0xc0050dc340, 0x1488a01, 0xc004de48b8)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/consumer_partition.go:580 +0xf4b
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).handleMessage(0xc004de4840, 0xc0050b1360, 0x1e71520, 0xc0050dc340)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/connection.go:682 +0x112
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).internalReceivedCommand(0xc004de4840, 0xc0050d56c0, 0x1e71520, 0xc0050dc340)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/connection.go:558 +0x287
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).run(0xc004de4840)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/connection.go:412 +0x3e2
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).start.func1(0xc004de4840)
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/connection.go:227 +0x81
created by github.com/apache/pulsar-client-go/pulsar/internal.(*connection).start
        /usr/local/app/go/pkg/mod/github.com/apache/pulsar-client-go@v0.7.0/pulsar/internal/connection.go:223 +0x56

```

--- a/pulsar/consumer_partition.go
+++ b/pulsar/consumer_partition.go
@@ -585,7 +585,7 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
 
 	for i := 0; i < numMsgs; i++ {
 		smm, payload, err := reader.ReadMessage()
-		if err != nil {
+		if err != nil || payload == nil {
 			pc.discardCorruptedMessage(pbMsgID, pb.CommandAck_BatchDeSerializeError)
 			return err
 		}

--- a/pulsar/internal/buffer.go
+++ b/pulsar/internal/buffer.go
@@ -19,6 +19,7 @@ package internal
 
 import (
 	"encoding/binary"
+	log "github.com/sirupsen/logrus"
 )
 
 // Buffer is a variable-sized buffer of bytes with Read and Write methods.
@@ -110,6 +111,11 @@ func (b *buffer) IsWritable() bool {
 }
 
 func (b *buffer) Read(size uint32) []byte {
+	// Check []byte slice size, avoid slice bounds out of range
+	if b.readerIdx+size > uint32(len(b.data)) {
+		log.Errorf("The input size [%d] > byte slice of data size [%d]", b.readerIdx+size, len(b.data))
+		return nil
+	}
 	res := b.data[b.readerIdx : b.readerIdx+size]
 	b.readerIdx += size
 	return res

--- a/pulsar/internal/buffer.go
+++ b/pulsar/internal/buffer.go
@@ -19,6 +19,7 @@ package internal
 
 import (
 	"encoding/binary"
+
 	log "github.com/sirupsen/logrus"
 )
 

